### PR TITLE
Dev loll

### PR DESCRIPTION
**Pull request by lollcat**
- Add training by ML for GMM and Many Well using exact samples. Set loss_type = target_forward_kl to use this in the config file. 
- Add rejection sampling for Double Well to obtain exact samples from the first dim

### CODE REVIEW

File: fab/core.py [hunk 1]
```diff
@@ -146,6 +149,11 @@ def fab_alpha_div_loss(self, batch_size: int) -> torch.Tensor:
         loss = self.fab_alpha_div_loss_inner(x_ais, log_w_ais)
         return loss
 
+    def target_forward_kl(self, batch_size: int) -> torch.Tensor:
```
Comment (lollcat) [inline fab/core.py:152]: Use "target_foward_kl" in the config file to use this training setting. Assumes the target has a .sample method

File: fab/target_distributions/double_well.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
+from typing import Optional, Dict
```
Comment (lollcat) [inline fab/target_distributions/double_well.py:1]: Separated out the Double well into its own file to clean things up

File: fab/utils/numerical.py [hunk 3]
```diff
@@ -22,12 +22,22 @@ def effective_sample_size(log_w: torch.Tensor, normalised=False):
...
+
 
-def quadratic_function(x: torch.Tensor):
+def quadratic_function(x: torch.Tensor, seed: int = 0):
+    torch.random.manual_seed(seed)
     # example function that we may want to calculate expectations over
     x_shift = 2*torch.randn(x.shape[-1]).to(x.device)
     A = 2*torch.rand((x.shape[-1], x.shape[-1])).to(x.device)
     b = torch.rand(x.shape[-1]).to(x.device)
+    torch.seed() # set back to random number
```
Comment (lollcat) [inline fab/utils/numerical.py:40]: Have to make sure that the randomness for the quadratic function is dealt with correctly so that it is the same for each evaluation.

File: examples/make_flow/__init__.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
 from examples.make_flow.make_realnvp_normflow import make_wrapped_normflowdist, make_normflow_model, make_normflow_snf_model
 from examples.make_flow.make_masked_affine_nflows import make_wrapped_nflows_dist
 
-from examples.make_flow.make_aldp_model import make_aldp_model
\ No newline at end of file
+# from examples.make_flow.make_aldp_model import make_aldp_model
```
Comment (lollcat) [inline examples/make_flow/__init__.py:4]: Adding the aldp stuff to the `__init__.py` files means that we have to install quite a few extra dependencies. I think we should make these denendencies optional so I commented this out. What do you think?

Comment (VincentStimper) [reply examples/make_flow/__init__.py:4]: I see that it is annoying for you to install them now when they are not needed, so this is fine for now.

**Review by VincentStimper — COMMENTED** [review]
Why did you add the models? I think they should not be part of the repository and might cause problems with the file size limit later on.

**Review by VincentStimper — APPROVED** [review]
